src/lib/OS/SyscallHandler.cc

```diff
@@ -1,15 +1,712 @@
 #include "simeng/OS/SyscallHandler.hh"
 
-#include "simeng/OS/SimOS.hh"
 namespace simeng {
 namespace OS {
 
-SyscallHandler::SyscallHandler(SimOS* os) : os_(os) {
+SyscallHandler::SyscallHandler(
+    const std::unordered_map<uint64_t, std::shared_ptr<Process>>& processes,
+    std::shared_ptr<simeng::memory::Mem> memory,
+    std::function<void(simeng::OS::SyscallResult)> returnSyscallResult,
+    std::function<uint64_t()> getSystemTime, VAddrTranslator vAddrTranslation,
+    mmapFileOnHost mmapHostFd)
+    : processes_(processes),
+      memory_(memory),
+      returnSyscallResult_(returnSyscallResult),
+      getSystemTime_(getSystemTime),
+      vAddrTranslation_(vAddrTranslation),
+      mmapHostFd_(mmapHostFd) {
   // Define vector of all currently supported special file paths & files.
   supportedSpecialFiles_.insert(
       supportedSpecialFiles_.end(),
       {"/proc/cpuinfo", "proc/stat", "/sys/devices/system/cpu",
        "/sys/devices/system/cpu/online", "core_id", "physical_package_id"});
+
+  resumeHandling_ = [this]() { handleSyscall(); };
+}
+
+void SyscallHandler::receiveSyscall(SyscallInfo info) {
+  syscallQueue_.push(info);
+}
+
+void SyscallHandler::tick() { resumeHandling_(); }
+
+void SyscallHandler::handleSyscall() {
+  if (syscallQueue_.empty()) return;
+
+  const SyscallInfo info = syscallQueue_.front();
+  ProcessStateChange stateChange;
+
+  switch (info.syscallId) {
+    case 29: {  // ioctl
+      int64_t fd = info.registerArguments[0].get<int64_t>();
+      uint64_t request = info.registerArguments[1].get<uint64_t>();
+      uint64_t argp = info.registerArguments[2].get<uint64_t>();
+
+      std::vector<char> out;
+      int64_t retval = ioctl(fd, request, out);
+
+      assert(out.size() < 256 && "large ioctl() output not implemented");
+      uint8_t outSize = static_cast<uint8_t>(out.size());
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {retval}};
+      stateChange.memoryAddresses.push_back({argp, outSize});
+      stateChange.memoryAddressValues.push_back(
+          RegisterValue(reinterpret_cast<const char*>(out.data()), outSize));
+      break;
+    }
+    case 46: {  // ftruncate
+      uint64_t fd = info.registerArguments[0].get<uint64_t>();
+      uint64_t length = info.registerArguments[1].get<uint64_t>();
+      stateChange = {
+          ChangeType::REPLACEMENT, {info.ret}, {ftruncate(fd, length)}};
+      break;
+    }
+    case 48: {  // faccessat
+      int64_t dfd = info.registerArguments[0].get<int64_t>();
+      uint64_t filenamePtr = info.registerArguments[1].get<uint64_t>();
+      int64_t mode = info.registerArguments[2].get<int64_t>();
+      // flag component not used, although function definition includes it
+      int64_t flag = 0;
+
+      char* filename = new char[PATH_MAX_LEN];
```

Comment:
To avoid using `new`, could this be changed to a std::array or std::vector where the underlying pointer could be retrieved when passing into faccessat?

Reply:
This applies throughout this file wherever `readStringThen` is used

Reply:
Made changes to all instances